File: devcamp2018/app/src/main/java/com/psato/devcamp/infrastructure/ProjectViewModelFactory.java
package com.psato.devcamp.infrastructure;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.v4.util.ArrayMap;

import com.psato.devcamp.di.component.ViewModelSubComponent;
import com.psato.devcamp.presentation.MVVM.QueryViewModelArc;
import com.psato.devcamp.presentation.home.HomeFragmentViewModel;

import java.util.Map;
import java.util.concurrent.Callable;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class ProjectViewModelFactory implements ViewModelProvider.Factory {
    private final ArrayMap<Class, Callable<? extends ViewModel>> creators;

    @Inject
    public ProjectViewModelFactory(ViewModelSubComponent viewModelSubComponent) {
        creators = new ArrayMap<>();

        // View models cannot be injected directly because they won't be bound to the owner's
        // view model scope.
        creators.put(HomeFragmentViewModel.class, () -> viewModelSubComponent.homeFragmentViewModel());
        creators.put(QueryViewModelArc.class, () -> viewModelSubComponent.queryViewModelArc());
    }

    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        Callable<? extends ViewModel> creator = creators.get(modelClass);
        if (creator == null) {
            for (Map.Entry<Class, Callable<? extends ViewModel>> entry : creators.entrySet()) {
                if (modelClass.isAssignableFrom(entry.getKey())) {
                    creator = entry.getValue();
                    break;
                }
            }
        }
        if (creator == null) {
            throw new IllegalArgumentException("Unknown model class " + modelClass);
        }
        try {
            return (T) creator.call();
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
}

File: devcamp2018/app/src/main/java/com/psato/devcamp/infrastructure/ProjectViewModelFactory.kt
package com.psato.devcamp.infrastructure

import android.arch.lifecycle.ViewModel
import android.arch.lifecycle.ViewModelProvider
import android.support.v4.util.ArrayMap

import com.psato.devcamp.di.component.ViewModelSubComponent
import com.psato.devcamp.presentation.search.QueryViewModelArc
import com.psato.devcamp.presentation.home.HomeFragmentViewModel

class ProjectViewModelFactory(viewModelSubComponent: ViewModelSubComponent) : ViewModelProvider.Factory {
    private val creators: ArrayMap<Class<*>,() ->ViewModel> = ArrayMap()

    init {
        creators[HomeFragmentViewModel::class.java] = { viewModelSubComponent.homeFragmentViewModel() }
        creators[QueryViewModelArc::class.java] = { viewModelSubComponent.queryViewModelArc() }
    }

    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        var creator: (() ->ViewModel)? = creators[modelClass]
        if (creator == null) {
            for ((key, value) in creators) {
                if (modelClass.isAssignableFrom(key)) {
                    creator = value
                    break
                }
            }
        }
        if (creator == null) {
            throw IllegalArgumentException("Unknown model class $modelClass")
        }
        try {
            return creator.invoke()as T
        } catch (e: Exception) {
            throw RuntimeException(e)
        }

    }
}